### ISSUE TITLE
event.stopPropagation() doesn't stop event propagating to view

### ISSUE THREAD

**Issue by georeith** [paperjs/paper.js#995]
Calling `event.stopPropagation()` does not stop an event propagating to the view.

```js
item.on('mousedown', function(event) {
    event.stopPropagation();
    console.log('item');
});
view.on('mousedown', function(event) {
   console.log('view');
});
```

Also the event on the view has the target set as the view, although the original target was the item, is this done on purpose?

**Edit:** The issue is here:

```js
function emitMouseEvent(obj, type, event, point, prevPoint, stopItem) {
	var target = obj,
		prevented = false,
		mouseEvent;

	function emit(obj, type) {
		if (obj.responds(type)) {
			if (!mouseEvent) {
				mouseEvent = new MouseEvent(type, event, point, target,
						prevPoint ? point.subtract(prevPoint) : null);
			}
			mouseEvent.id = point;
			if (obj.emit(type, mouseEvent)) {
				called = true;
				if (mouseEvent.prevented)
					prevented = true;
				if (mouseEvent.stopped)
					return true;
			}
		} else {
			var fallback = fallbacks[type];
			if (fallback)
				return emit(obj, fallback);
		}
	}

	while (obj && obj !== stopItem) {
		if (emit(obj, type))
			break;
		obj = obj._parent;
	}
	return prevented;
}

function emitMouseEvents(view, item, type, event, point, prevPoint) {
	view._project.removeOn(type);
	called = false;
	return (dragItem && emitMouseEvent(dragItem, type, event, point,
				prevPoint)
		|| item && item !== dragItem && !item.isDescendant(dragItem)
			&& emitMouseEvent(item, fallbacks[type] || type, event, point,
				prevPoint, dragItem)
		|| emitMouseEvent(view, type, event, point, prevPoint));
}

```

Propagation to view seems to be handled by `emitMouseEvents` if the first two OR statements are false. However propagation detection is in `emitMouseEvent` and uses a closured `MouseEvent` so the view receives a different one with the wrong target and without the `stopped` property.

**Comment by lehni** [comment]
I agree that this is a bug. Regarding the wrong target on view, could you clarify what you expect to receive?

**Comment by georeith** [comment]
@lehni I would expect to see the original Item you clicked, so it can still be view sometimes (if it isn't a propagated event), the same way it is supposed to behave in your code but doesn't because its outside of the closure.

**Comment by lehni** [comment]
At the moment, for reasons of performance I only hit-test for items if there are any item events installed... This would break that. It's intentional, not a problem of closures. The event on view never receives another target than the view, I believe...

**Comment by georeith** [comment]
@lehni Sorry I should of said the highest one with a listener.

**Comment by georeith** [comment]
The issue is that `emitMouseEvent(view, type, event, point, prevPoint));` handled the propagation when it should of been handled by the original `emitMouseEvent` and still contain a reference to the original target.

It should be the same event that propagates and an events target never changes is the real point.

**Comment by georeith** [comment]
That should handle the propagation:

```js
while (obj && obj !== stopItem) {
        if (emit(obj, type))
            break;
        obj = obj._parent || obj.view;
}
```

But I'm not entirely sure what's going on in the big boolean return statement in `emitMouseEvents` I think I'd do more harm than good touching that as I'm not sure what issues you are fighting against.

**Comment by lehni** [comment]
I have already fixed your issue, I just haven't pushed it yet. There is nothing wrong with the propagation the way it's currently written, there was just a mix-up between preventing the default and stopping propagation. There are good reasons for the current separation of items and view (handled by that boolean statement), the lack of a need to hit-test for items when all you want is an event listener on the view being one of them.

There is no should or shouldn't, really. This is not the DOM, we can make up our own rules. People expect all kinds of things, though. There is never going to be one solution that makes everybody happy.

**Comment by lehni** [comment]
PS: Your proposal only works if you actually hit an item. If there is none, then the view would not be receiving the event either.

**Comment by lehni** [comment]
PPS: If you are curious about the code, I'd recommend looking a the sources, which are full of explaining comments, e.g.: https://github.com/paperjs/paper.js/blob/develop/src/view/View.js#L1197

**Comment by lehni** [comment]
Reopening, since there are open questions still:
- Should the event on view contain the hit-item as target as well?
- event.preventDefault() in mousedown does not prevent mousedrag events on the view, but does so on items.

**Comment by georeith** [comment]
@lehni Internally it makes sense to differentiate the view and items but as a user a common interface would be far more useful I think.

Ah I was looking at the `prebuilt/module` source and it had comments stripped. Didn't realise it had all that.

What benefit does `target` provide if it is the same thing you clicked on? You could closure that into your event callback. It does solve the currently impossible that is an event propagates to view and I want to know A if the event was propagated and B where it began.

**Comment by lehni** [comment]
I think you should give the code another go since my fix above. It does exactly what you expect it to. The only difference right now is target.

**Comment by lehni** [comment]
I don't want to run hit-tests that people might not even use, which is why there currently is the optimization that counts the amount of mouse events installed on items, and only runs hit-tests when there are items that expect events to be called on them. This is a good thing, and works really well (see View._itemEvents). Your proposal would make this useless, and would result in hit-tests being run all the time. I could add a getter to the event object that is passed to events on view for target that would run the hit-test once it's requested by the user. I considered that before but rejected it as it will add more code. All that said to explain that there is a good reason why things are implemented the way they currently are.

**Comment by georeith** [comment]
@lehni I have just isntalled it and it is working great. The `target` isn't something I am using now (I used to for the reason it wasn't working correctly).

I don't understand why my proposal needs you to run additional hit-tests? All I'm proposing is that if an event was propagated by something else then keep that as the target even when it hits view. Right now view always receives view as the target even if it wasn't the original receiver.

**Comment by lehni** [comment]
For the simple reason that if I am not installing any events on items and only on the view, *no* hit-tests are run at all at the moment, the actual item target doesn't have to be determined, and that's a good thing. If you want the event target, you'd have to run a hit-test to find out. This could be masked behind a `event.target` getter if required, but so far there was no need for it.

This is what I've been trying to explain in all my comments here: This is the reason why things are implemented the way they are, and that's by design, and not a bug.

**Comment by georeith** [comment]
@lehni but thats fine if there are no events on any items then the target would be the view because the event didn't propagate. I think you are misunderstanding what I'm saying.

**Comment by lehni** [comment]
No, I don't think so... But I think it's wrong that you get a different target on view events based on wether you have item events installed or not.

**Comment by georeith** [comment]
@lehni You just get whatever triggered the first emit. No need for any additional hit-tests. If you can call `obj.emit` then you already know what the object is.

**Comment by lehni** [comment]
You are wrong. This has nothing to do with closures. You make wrong assumptions about the code I've written, and I'm getting tired with trying to explain it.

**Comment by georeith** [comment]
@lehni I'm not assuming.`emit` contains a closured `obj` from its wrapper `emitMouseEvent` anything it emits will emit a mouseEvent with the target as `obj` no?

Right now it doesn't propagate to the view... so when you call the wrapper again based off of the same initial event the view only ever receives the view as the target. Why should the view be any different than the other events? It doesn't make any sense.

**Comment by lehni** [comment]
There really is no need to explain the code to me, I've written it, I know what it does : )
What I am saying is is this:
- When you have no events installed on any of the items in the project, then none of the event counters is increased in `View#_itemEvents`. (Perhaps study the code that relates to that structure)
- In `_handleMouseEvent()`, `View#_itemEvents` is used to determine if a hit-test should be run (stored in `hitItems`). Only if any of the times is expecting an event, the hit-test is run. If none is expecting an event, it doesn't mean there is no item under the mouse. It just means no events expects the hit-test result

From here the rest should make sense.

**Comment by georeith** [comment]
@lehni but I'm not talking about whats under the mouse:

- If there is a square under the mouse and it has no listener but the view does then the target is the view.
- If there is a square under the mouse and it has a listener then it is the target.

The target is the first item, you've already done the hitTest because you've already called `obj.emit` on it. All I'm saying is that you extend how it currently works for child -> parent to treat view like a parent and not its own separate event.

The whole issue is contained with `emitMouseEvents`.

**Comment by lehni** [comment]
And I'm saying that the target shouldn't be different base on wether an item has an event installed or not. We're looping.

**Comment by georeith** [comment]
@lehni But what purpose does always having the view as a target serve? Why do Items get different targets. It's nothing to do with whether events are installed but whether the event propagated.

**Comment by lehni** [comment]
An API that changes its behavior based on wether some / any item (not even the item under the mouse, necessarily) has an event installed or not, is erratic and to be avoided at all cost. But that's what your proposal would do. As soon as *any* item has an event, the hit-test is run and the target item is found, wether it has the event installed or not.

The discussion wether view as the target of view events makes sense or not is a separate one, and I said that there is a way to make target work differently (through the accessor), but I didn't see the need for it and decided it didn't merit the added code. If you want, you can just think of view events not having a target currently, and be fine with that. Or you can argue it should have a target and it should be an item or the view if there is none, and that would require new code, as your proposal wouldn't solve that either.

**Comment by georeith** [comment]
@lehni it already changes its behaviour... the current behaviour is the erratic one.

HTML events are erratic? They don't do that.

**Comment by lehni** [comment]
Here we loop again.

**Comment by georeith** [comment]
@lehni Can you tell me the definition of `target`? My one version is well defined the other is erratic. Please describe it and hopefully you will see.

**Comment by lehni** [comment]
Dude, whatever.

**Comment by elliotbonneville** [comment]
I think what @georeith is saying is that an event should retain the original `target` even after bubbling up through the view. `target` is already determined and stored -- it's just dropped when the event hits the view. It seems to make more sense to keep that information. Do you have a reason not to?

**Comment by lehni** [comment]
I thought I explained this enough times now, but I have a feeling my explanations are being ignored, which is rather irritating, as explaining things takes time away from other work that I could focus on other things instead. Also, it is probably never a good idea to ignore the explanations of the a maintainer of a piece of open source software that you rely on, and I can't imagine a situation where this is ever the right tone:

> Please describe it and hopefully you will see.

But fine, I'll explain one last time why I think the proposal in its current is erratic:

1)
	- A view with one item in the center
	- The item has no event handler installed
	- The view as a mousedown handler installed
	- User clicks in the center of the view, view receives itself as target, because no hit-test is executed (non of the items have events)
2)
	- The same scenario as 1), but with these changes:
	- The item has an mousedown handler installed that doesn't stop propagation 
	- User clicks in the center of the view, hit-test is run because items have event handlers, center item is found and receives the event, the event bubbles up, view receives the center item as target
	-> Behavior of view event changes based on the presence of one item that receives events
3)
	- The same scenario as 1), but with these changes:
	- A second item is introduced at the border of the view
	- This second item has an mousedown handler installed
	- User clicks in the center of the view, hit-test is run because items have event handlers, center item is found but does not receive the event because it has no handlers installed, the event bubbles up, view receives the center item as target
	-> Behavior of view event changes based on the presence of one outside the reach of the mouse item that receives events

This is what I think is an erratic API and to be avoided. Feel free to proof otherwise, but please note that I enver said that I think it is *good* that the current view handlers only receive the view as targets. This was simply the left-over of an optimization, and I already made suggestions s to how this could be fixed proberly, in a way that is not causing the issues outlined above.

**Comment by elliotbonneville** [comment]
Thank you for taking the time to write another verbose explanation, @lehni. I'm sorry if you felt your earlier explanations were being ignored -- I was just having difficulty understanding them. I'm new to the codebase. : )

Now, the first two scenarios make sense, but I'm curious as to why in the third scenario you say that `target` would be the center item. It seems to me that in this case, the view should still receive itself as the target, because there are no valid `target`s (items that have a(n) event listener(s) installed) that would pass the hit-test, right? Or is there nothing in place right now that filters the results of the hit-test to check for event listeners?

**Comment by lehni** [comment]
@elliotbonneville I'm not explaining how it should work, I'm just trying to explain how the change proposed by @georeith would work, and why I think that would be problematic. I don't think we need a new issue.

**Comment by lehni** [comment]
Also, I don't think in HTML targets are only items that have events installed. I think it really is the item under the pointer.

**Comment by lehni** [comment]
To clarify a possible confusion here: In HTML, when I install a mousedown event on the document, `event.target` is not pointing to the item that has the handler installed, it's the element that's actually under the mouse when I click. I am not sure where assumption comes from that in paper.js, `event.target` should point to the item with the handler installed:

https://jsfiddle.net/kwhu7LkL/

**Comment by georeith** [comment]
@lehni Either version is good. That is just a suggestion so that an additional hit-test is not required, my concern is that `event.target` being the item that is currently being dispatched isn't useful and that as it currently stands it follows no clear pattern.

The code I provided was meant more as an explanation of what I meant. I'm sure it's riddled with bugs it was made as a quick simplification of what I meant.

Please don't take our discussion personally, I apologise if you feel that way my only intentions are to maintain the integrity of the library.

**Comment by lehni** [comment]
@georeith I'm not taking it personal, I'm just getting a bit frustrated as you still somehow seem oblivious to what I have been trying to explain for the past dozen of comments or so, and I am about to run out of ways to explain it: If `event.target` is to be consistent in the way it is in HTML (or any other way really), then additional hit-tests are required in the situations where no item has any event handlers installed. I've tried to explain that in so many ways, I've provided example scenarios illustrating what I am trying to say, and I've provided an HTML example to compare it with, but you seem to be ignoring all this and stating that both versions are fine. I would say the opposite: Neither version is fine, and if the consistency of `event.target` is important then there is a solution, but neither the current state nor your proposal will solve that.

I have a solution in my head already, so I'll stop discussing this further and will spend my time implementing it instead.

But I'd recommend to consider your style of discussing these things on Github. I know what closures are and how they work. I know what the code does that I wrote myself. To suggest otherwise feels condescending and doesn't set a good tone in an open-source project, where the participation is based on generosity, passion in the subject and a willingness to donate time to a shared goal.

**Comment by georeith** [comment]
@lehni I realise that and was offering an alternate solution to the HTML spec. If you don't want to implement it that's fine, it's not been clear to me that you don't want that definition for target so much as you think I'm suggesting we implement the HTML spec.

I think you are being unfair and some of your comments come across as irate where no offence has been intended. I'm not suggesting you don't know how things work, I wouldn't even be discussing this with you if I didn't value your opinion. There is no way for me to respond to something like this if I disagree that you wouldn't take offensively:

> You are wrong. This has nothing to do with closures. You make wrong assumptions about the code I've written, and I'm getting tired of trying to explain it to you.

Please understand I am also just trying to help and an open-source community is about discussion and having multiple eyes on something.

Regardless this is becoming too personal for my liking and I would ask that we can just pretend this is 3rd comment in this issue and go back to our happy selves?

**Comment by lehni** [comment]
Yes I agree, this spiraled out of control. The whole event handling code is a complex beast and I've spent far too much time getting to where it is right now, which is a full rewrite of the previous code. I'm very reluctant to making changes on it, also because it is harder to unit-test than the rest of the library... Happy to go back to this.

![keep-on-trucking](https://cloud.githubusercontent.com/assets/213275/13373609/9bc0fa22-dd6c-11e5-901e-c00f46c0b097.jpg)

**Comment by georeith** [comment]
@lehni Totally understand, I really value that. Thanks dude.

**Comment by lehni** [comment]
@georeith I finally managed to find a nice fix for this issue, that should make us both happy : )

**Comment by georeith** [comment]
@lehni Awesome, just reading the code am I right in thinking it returns the item the event triggered on or in the case that it was a listener on the view runs a hitTest and gets the highest item under the mouse?

**Comment by lehni** [comment]
Yes that's what it's doing... I'm just now realizing that this is not how Event#target is defined in HTML land. argh. I'm giving up : ) https://developer.mozilla.org/en-US/docs/Web/API/Event/target

**Comment by lehni** [comment]
I thought it was weird that when you install a listener on view, target would be different based on whether you also installed a listener on an item that the user clicks on or not.

**Comment by georeith** [comment]
You mean if I have a listener on the view and on the item that I click? I think HTML has the target be the same throughout... whichever first caught the event, then `currentTarget` is the current item that's receiving the event.

So in that case in HTML I believe it would be:

trigger event on item -> target = item, currentTarget = item
propagate to view -> target = item, currentTarget = view

**Comment by lehni** [comment]
Oh cool, so we're consistent.

**Comment by georeith** [comment]
But as you said I don't think we have to match the HTML spec, although I guess it's nice for interoperability. The main thing is just being able to tell if event was propagated and what it was propagated from.

**Comment by georeith** [comment]
I believe that's what this means:

> event . target
> Returns the object to which event is dispatched.
> 
> event . currentTarget
> Returns the object whose event listener's callback is currently being invoked.

Dispatched being the item which initiated the event. And currentTarget being a bit more self explanatory. Find the spec hard to read as you often have to follow their other definitions, but I would assume thats what they mean by dispatched here.

**Comment by lehni** [comment]
How do you feel about the current implementation? I think it's consistent.

**Comment by georeith** [comment]
@lehni I believe we differ from the spec slightly. Take this example:

https://jsfiddle.net/gv0hzrtf/

If you click on the red they both receive it as the target and then themselves as the `currentTarget` even though it has no event listener.

I like both ways the current kind because you can tell when something is propagated (target is not itself), but thinking about the HTML way it allows you to put a listener on the parent (in our case the view) and use it as a listener for all elements (delegated events). With the implementation above you would need to this manually with a hitTest to be sure. Not sure which is more useful now I think about.

**Comment by lehni** [comment]
Isn't `currentTarget` the same as `this` inside the callback?

**Comment by georeith** [comment]
Yeah it is, nothing major just context can always be changed, especially with ES6 arrow methods and lexical scoping would be nice to have consistent access to that value. 99% of the time though `this` is fine.

**Comment by lehni** [comment]
We could probably add it in a really simple way, through `Emitter#emit()`. If there's an event object, it could set `#currentTarget` to `this` before emitting the event...

**Comment by lehni** [comment]
If that's too extreme (probably), then maybe the `Event` classes that should behave this way could define a "private" property called `#_updateTarget`, and if that's defined, it would do the above.

**Comment by georeith** [comment]
Sounds good.

One question now I've had time to think about it. Are we being consistent with the hitTest as sometimes its the `this` value of the first triggered listener and other times its the highest item under the mouse no? I might just be misunderstanding the possible values of `Event#target` though.

Perhaps we should just set `Event#target` to whatever the first `currentTarget` is and ensure it is always the same throughout. If its the view its the view, if it fires on a object before propagating to view then it is that, assuming we do child -> parent propagation and not just view (haven't actually tested that).

E.g., 
http://sketch.paperjs.org/#S/nZKxbsMgEEB/BbE4kRCyGV11ytC9GWMPFF9sZHpYmCRKIv97wbFrR0qXMqDjne7ugbhTlN9Ac7pvwauGMqpsFc9n6YgD5SXWBjLyThAuZN/IDvjnjDf3AklYndXoc3JIGUlL9mC9vkFAWRpg2GZ81MbsrLEuJ4m6SkwmHi0C+p0oYmLYvhVY4JOK+LeKiCritYqD6i+TLHkSqZ09dZPDR4znyVPdmJ97qUabygGG6csF2Opdy6W3hgu3uDNataH92GZ1XkpeQbGCxxMqry1u4Azot2TSUxZ7a4AbWz8y3EtXg+fRmxHf6H4Mo80QvsGXA9mOr9nT/FAOPw==

click on rectangle1:
   1. target: rectangle1, currentTarget: rectangle1
   2. target: rectangle1, currentTarget: group
   3. target: rectangle1, currentTarget: view

click on rectangle2:
   1. target: rectangle2, currentTarget: rectangle2
   2. target: rectangle2, currentTarget: group
   3. target: rectangle2, currentTarget: view

click on group:
  1. target: group, currentTarget: group
  2. target: group, currentTarget: view

click on view:
  1. target: view, currentTarget: view

Unless I'm misunderstanding `Event#target` there and it is always the highest visible item like the DOM (if so ignore this I'm being a numpty). This lets you do everything, determine that the event is propagated and from where, and if you need to delegate events just `hitTest` the item with the `event.point` to get the child it hit.

**Comment by lehni** [comment]
Yeah that's what I was wondering about too : ) So this whole `hitTest()` / `getTarget()` magic wasn't even necessary. Oh well. : )

**Comment by georeith** [comment]
Ha sorry I can't really remember what we were going over before it might have already been suggested, was just running it over my head last night and seems like the most elegant without degrading performance. Also I messed up the naming a bit in that sketch this is a better one:

http://sketch.paperjs.org/#S/nZKxbsMgEIZfBbE4kZDlMLrqlKF7M8YeKL7YyOTOwiRRG/ndC44tO1K6lAFx34n/PiHuHNUZeM4PLXjdcME1VbG+KsccaK+wtrBj7wzhxg6N6iD9nPHmXiALqyODPmfHTLCsFA/Wmx8IaJcFGLYZn4y1e7Lkcpbob4XJxKNFQMvE2Bi2bwUW+KQi/60io4p8reKg+stEJk8itaNLNzl8xPM8ebo39ucs3RhbOcAwfckTq3ctl2wDt5Rwb41uQ/wYs6qXK6+gXMHTBbU3hBu4Avotm/Q0YU8WUkv1o5N65WrwafQWzDemH4/RZgjf4MuBasfX7Hl+LIdf

**Comment by lehni** [comment]
Yeah I think I had a misconception about what `#target` should return, hence our endless disagreements and looping earlier in the year, probably. Apologies for that. I will revert the `hitTest()` magic.

**Comment by georeith** [comment]
@lehni Not at all. I think we both just had a stressful day and felt we were right, I know you were fed up with the events just having refactored them and rightly so, I know the feeling.

**Comment by lehni** [comment]
@georeith yeah I do remember some preexisting levels of stress that day, unrelated : )

I do like where we ended up here, code looks good! I wasn't keen on the `hitTest()` hack.

It would be good to have unit tests in place for all this mouse stuff... I guess I could just emulate prerecorded mouse sequences by the use of `dispatchEvent()`?

**Comment by elliotbonneville** [comment]
That's what we've settled on doing with our unit tests. Here's what we've been using to dispatch events:

```
const scope = new paper.PaperScope();
const document = scope.document;

export const mouse = {
    move(x, y) {
        document.dispatchEvent(new window.MouseEvent('mousemove', {
            clientX: x,
            clientY: y,
        }));
    },

    down(x, y) {
        scope.project.view.element.dispatchEvent(new window.MouseEvent('mousedown', {
            clientX: x,
            clientY: y,
        }));
    },

    up(x, y) {
        document.dispatchEvent(new window.MouseEvent('mouseup', {
            clientX: x,
            clientY: y,
        }));
    },
};
```

**Comment by lehni** [comment]
@elliotbonneville that's great! We should do this too. What are you testing for this way?

**Comment by lehni** [comment]
@georeith now with `event.currentTarget`and docs.

**Comment by lehni** [comment]
@georeith I've been juggling too many things this week, didn't realize we made a wrong assumption, and this was masked by the accidental removal of a code optimization that only executes the hit-test when it's sure to be needed. I've added this optimization back now (ab24f92373245754621bb156e340a79b894853b7), and it shows that we do need the `hitTest()` getter magic. See this example:

https://jsfiddle.net/lehni/ou13xkzs/

The event is installed on the outmost `<div>` only, yet `event.target` points at whatever item is on top of the stack where you click, wether it subscribes to the event or not. This is what I wanted to achieve with this additional code, and also what I pointed out in your suggestion earlier this year as a short-coming. 

I hope we can finally clear this confusion up : ) I think we have two options: Keep the code simple and always run a hit-test, wether we need it or not (slow), or bring back my `getTarget()` getter that uses the previously run hit-test result if it exists or performs a hit-test if it's the first time it's requested.

**Comment by georeith** [comment]
@lehni Indeed that is the HTML way. I was thinking we would only do top of stack if it subscribes to the event (which deviates from HTML but saves additional hit-testing) leaving it up to the developer to do it or not.

So what I was thinking was the first emitted callback sets `event#target` to its owner. And any additional propagation has the same value just setting `event#currentTarget` to its owner.

Although I'm happy with if we do match the HTML spec too. Just worried it will make mousemove listeners slow for scenes with a large number of items. Not sure how optimised hit-test itself is. Under what circumstances does the hit-test already exist? Originally I imagined you ran a hit-test only on items with listeners for that event.

**Comment by lehni** [comment]
The way I had written is, the hit-test is only run when it's directly required (e.g. because of `click`, `mouseenter`, `mouseleave` events), or when `event.target` is first accessed. And per handling of event, the hit-test is never run more than once. So if you don't use `event.target`, there is never any impact on performance. And if you do, then it is necessary to behave the same as HTML, which I think is the only logic way now really...

**Comment by georeith** [comment]
@lehni Ah ok I like that it is in a getter on `event.target`. Although is it not possible that could be messed up? As in my event callback I move an item over `event#point` and then access `event#target` would it not wrongly report that item as being the `event#target`?

**Comment by lehni** [comment]
I guess it would, but I guess we could live with that? : )

**Comment by georeith** [comment]
@lehni Hmmm I'm just worried it would create more issues. Its a bit of a strange side effect to include in code, having to remember to read `event.target` before modifying anything.

**Comment by lehni** [comment]
Yeah I agree.

**Comment by lehni** [comment]
Always run the test?

**Comment by georeith** [comment]
@lehni I guess we have to to match HTML spec.

**Comment by lehni** [comment]
@georeith matching the spec means always running the hit-test. We could also activate it by default but offer a flat that turns it off?

**Comment by lehni** [comment]
What I mean is: There could be a need for no "correct" item-level event support. It could be nice to be able to install events on the view only, and not have paper.js take care of event#target, for performance reasons.

**Comment by georeith** [comment]
@lehni Sounds good, I would just return `null` or throw an error for the target if you switch it off instead of returning a possibly incorrect value.
